src/utils/hierarchical_model.py

```diff
@@ -0,0 +1,181 @@
+# Type hint
+from numpy import ndarray
+
+# Necessery libraries 
+import pandas as pd
+import numpy as np
+
+# Hierarchical clustering
+from scipy.cluster.hierarchy import dendrogram, linkage, fclusterdata
+
+# visulaization
+import matplotlib.pyplot as plt
+
+# Dimensionality reduction
+from sklearn.decomposition import PCA
+
+
+# Define a list of colors to choose from
+colors_code = ['#609EA2', '#AA5656', '#F2921D', "#4E6C50"]
+
+
+def get_linkage(data: pd.DataFrame, method_type='ward', distance_metric='euclidean', optimal_ordering=False) -> ndarray:
+    """Calculates hierarchical clustering linkage matrix for a given dataset.
+
+    Args:
+        data (pandas.DataFrame): The dataset to cluster.
+        method_type (str, optional): The clustering algorithm to use. Default is 'ward'.
+        distance_metric (str, optional): The distance metric to use. Default is 'euclidean'.
+        optimal_ordering (bool, optional): Whether to compute the optimal leaf ordering for the dendrogram. Default is False.
+
+    Returns:
+        numpy.ndarray: The hierarchical clustering linkage matrix.
+    """
+    return linkage(data, method=method_type, metric=distance_metric, optimal_ordering=optimal_ordering)
```

Comment:
no need for a function to do this

Reply:
I use this function in several places in the notebooks.